polygeist/lib/polygeist/Passes/ConvertToLLVMABI.cpp

```diff
@@ -87,22 +92,69 @@ class GPUFuncLowering : public OpRewritePattern<gpu::GPUFuncOp> {
   }
 };
 
+class GPUReturnLowering : public OpRewritePattern<gpu::ReturnOp> {
+public:
+  using OpRewritePattern<gpu::ReturnOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(gpu::ReturnOp op,
+                                PatternRewriter &rewriter) const override {
+    LLVM_DEBUG(llvm::dbgs()
+               << "ConvertToLLVMABIPass: GPUReturnLowering: " << op << "\n");
+
+    // Notify MLIR we're updating the function in place
+    rewriter.startRootUpdate(op);
+
+    bool Changed = false;
+    for (const auto &en : llvm::enumerate(op.getOperands())) {
+      auto MT = en.value().getType().dyn_cast<MemRefType>();
+      if (!MT)
+        continue;
+
+      Type PT = LLVM::LLVMPointerType::get(MT.getElementType(),
+                                           MT.getMemorySpaceAsInt());
+      auto memref2Ptr = rewriter.create<polygeist::Memref2PointerOp>(
+          en.value().getLoc(), PT, en.value());
+      op.setOperand(en.index(), memref2Ptr);
+      Changed = true;
+    }
+    LLVM_DEBUG({
+      if (Changed)
+        llvm::dbgs() << "  New ReturnOp: " << op << "\n";
+      else
+        llvm::dbgs() << " unchanged\n";
```

Comment:
[nit]: Suggest to remove and print only the operands that changed.

Reply:
prefer to keep it, as it is not for each operand, but only for a return op.